Feature/multiple hierarchies

havelaer opened the pull request:
Issue: #2451 

## What I did

Extended the hierarchy functionality to enable multiple hierarchies next to the main (nameless hierarchy).
I added an extra rootSeparator (not sure about the name though):
```
setOptions({
  // ... 
  hierarchySeparator: /\./,
  rootSeparator: /\|/, // default is false
});
```

**UPDATED**: `rootSeparator` is renamed to `hierarchyRootSeparator` with default value `null`

Extending the API as such:
```
storiesOf('Addons|Addon Info.React Docgen', module)
  .add(          ^
  ...
```

Works backwards compatible with or without setting the hierarchySeparator.

Documentation is updated in addons/options/README.md.

Allowing multiple hierarchies, see issue #2451 for screenshot.

## How to test

I added the rootSeparator to the kitchen sink example and added the "|" separator to the addon stories. 

See cra-kitchen-sink for an example.

The tests are fixed and extended. Though 1 test is failing, but that it is failing on master as well:
```
FAIL  examples/cra-kitchen-sink/src/storyshots.test.js
  ● Test suite failed to run

    TypeError: (0 , _react3.storiesOf)(...).add(...).add(...).add(...).addWithInfo is not a function
```

havelaer committed:
enable multiple hierarchies

havelaer committed:
extend kitchen sink example with multiple hierarchies

havelaer committed:
extend options addon to enable rootSeparator

havelaer committed:
updated documentation and rootSeparator default null

Hypnosphi commented on examples/cra-kitchen-sink/src/stories/addon-a11y.stories.js:
```diff
@@ -6,7 +6,7 @@ import BaseButton from '../components/BaseButton';
 const text = 'Testing the a11y addon';
 
-storiesOf('Addon a11y', module)
+storiesOf('Addons|Addon a11y', module)
```
Looks like now we can drop the "Addon" word everywhere:
`'Addons|A11y'`

Hypnosphi commented:
> 1 test is failing, but that it is failing on master as well

That's weird. Did you run `yarn bootstrap --core`?

Hypnosphi commented:
Can you please resolve the conflicts? They are mostly caused by renaming "Left panel" to "Stories panel"

havelaer committed:
omit Addon prefix from addon stories

havelaer committed:
fix test after rebase on release/3.3 branch

Havelaer commented:
@Hypnosphi Thanks for reviewing! 
The branch was based on master. I've rebased the branch on latest release/3.3 and fixed the conflicts. 
yarn bootstrap reset everything and all tests are passing!
I've also omitted 'Addon' in the addon story names.

igor-dv commented on lib/ui/src/modules/ui/containers/stories_panel.js:
```diff
@@ -25,11 +33,13 @@ export const mapper = (state, props, { actions }) => {
-  const selectedHierarchy = resolveStoryHierarchy(selectedKind, hierarchySeparator);
+  const storiesHierarchies = createHierarchies(filteredStories);
+
+  const [, storyName] = resolveStoryRoots(selectedKind, rootSeparator);
```
I don't understand why exporting the `resolveStoryRoots` ?  You can hide it under the `resolveStoryHierarchy` method.

Havelaer replied:
`resolveStoryRoots` is exported for 2 reasons:
1) It's also needed in the stories_panel container.
2) It's has separate unit tests.

Btw I renamed `resolveStoryRoots` to `resolveStoryHierarchyRoots`.

igor-dv commented on lib/ui/src/modules/ui/libs/hierarchy.js:
```diff
@@ -46,24 +46,47 @@ function fillHierarchy(namespaces, hierarchy, story) {
 
-export function createHierarchy(stories) {
-  const hierarchyRoot = {
+export function createHierarchyRoot(name = '') {
```
Looks like you export this function only for tests, IMO it's not a good thing.

Hypnosphi replied:
It's not the main js file, so I think it's Ok to have test-only exports for more granular unit tests

igor-dv replied:
👌

Hypnosphi commented on lib/ui/src/modules/ui/components/stories_panel/sub_header.js:
```diff
@@ -0,0 +1,36 @@
+const headingStyle = {
+  ...baseFonts,
+  textTransform: 'uppercase',
+  // letterSpacing: '1.5px',
```
Looks like it's worth uncommenting. Uppercase text is more readable when it has some letter-spacing

Havelaer replied:
Changed it to letterSpacing: '1.2px'

Hypnosphi commented:
Please run `yarn test --core --update` to update test snapshots

igor-dv commented on lib/ui/src/modules/ui/libs/hierarchy.js:
```diff
@@ -46,24 +46,47 @@ function fillHierarchy(namespaces, hierarchy, story) {
+
+export function resolveStoryRoots(storyName = '', rootSeparator) {
+  if (!rootSeparator) {
+    return ['', storyName];
```
I think in this method better return an object, instead of relying on indices in the array.

Havelaer replied:
Agreed. Changed it, so it returns an object.

Btw I renamed resolveStoryRoots to resolveStoryHierarchyRoots.

Hypnosphi commented:
[Live example](https://deploy-preview-2452--storybooks-cra.netlify.com/?selectedKind=Welcome&selectedStory=to%20Storybook&full=0&addons=1&stories=1&panelRight=1&addonPanel=storybook%2Factions%2Factions-panel) for reference

Looks like some of the stories still have wrong kinds:
![image](https://user-images.githubusercontent.com/6651625/33796097-bb33958c-dcf6-11e7-8f9a-54cae9ca72d4.png)

Hypnosphi commented:
Root title is hidden when using search. Is it on purpose?
![image](https://user-images.githubusercontent.com/6651625/33796112-fb56b004-dcf6-11e7-8e99-7d406ae9aa3c.png)

igor-dv commented on lib/ui/src/modules/ui/libs/hierarchy.js:
```diff
@@ -46,24 +46,47 @@ function fillHierarchy(namespaces, hierarchy, story) {
   }
 
-  return hierarchyRoot;
+  return Object.keys(rootMap).map(rootName => rootMap[rootName]);
```
`Object.values(rootMap)` ?

Havelaer replied:
Polyfills take care of `Object.values`?

Code is changed a bit at those lines though: I had to make sure that the main hierarchy is always the first in the array, so I didn't have use for `Object.values` anymore.

Which triggers the next question:
What could we do with the sorting of the other hierarchies in the panel? Now it respects the order in which the stories are loaded.

havelaer committed:
added missing addons to addon hierarchy

igor-dv commented on lib/ui/src/modules/ui/components/stories_panel/sub_header.js:
```diff
@@ -0,0 +1,36 @@
+);
+
+SubHeader.defaultProps = {
+  name: '',
```
what do you think about changing `name` to `children` ? I think it would be nicer for use. And I think it should be `required`.

Havelaer replied:
Agreed! Initially copied stories_panel/header.js props interface. But I prefer children prop in such situations. So changed it.

igor-dv commented on lib/ui/src/modules/ui/components/stories_panel/index.js:
```diff
@@ -50,17 +50,23 @@ class StoriesPanel extends Component {
-          {hierarchyContainsStories(storiesHierarchy) ? (
-            <Stories {...pick(this.props, storyProps)} />
-          ) : null}
+          {storiesHierarchies.map(
```
worth extracting it to function already.

Havelaer replied:
extracted the map it to separate render method.
extracted the sub header to stories_tree/ as `tree_header`

havelaer committed:
restored letterspacing sub header

igor-dv reviewed (changes requested):
1. Very like this ability. Would like to use for separating core component in my project.
2. I didn't check the tests yet 
3. I think overloading the storyKind to have more special syntax is a bit wrong direction, but currently, we have nothing to do with it. (maybe create some kind of URI syntax for story) @ndelangen , you have plans for refactoring, WDYT?

havelaer committed:
updated snapshots

havelaer committed:
change rootSeparator to hierarchyRootSeparator

havelaer committed:
sub header text as required children

havelaer committed:
refactor stories panel

havelaer committed:
refactor and fix hierarchy helpers

Havelaer commented:
> Looks like some of the stories still have wrong kinds:

Fixed!

> Root title is hidden when using search. Is it on purpose?

Nice catch! Fixed!

Havelaer commented on addons/options/README.md:
```diff
@@ -90,13 +90,19 @@ setOptions({
+   *   /\|/ - split by `|`
+   * @type {Regex}
+   */
+  rootSeparator: null,
```
done!

viddo replied:
Could the PR description be updated to have it in the example there? Didn't work on the first try and had to search around to discover this.

Found my way here through the [3.4.0-alpha0 changelog entry](https://github.com/storybooks/storybook/blob/next/CHANGELOG.md#340-alpha0)

Havelaer commented on addons/options/src/preview/index.js:
```diff
@@ -23,16 +27,17 @@ export function setOptions(newOptions) {
+  // override the hierarchySeparator or rootSeparator if the prop is missing
+  const options = {
+    ...newOptions,
+    ...(hasOwnProp(newOptions, 'hierarchySeparator')
```
Done! Called it withRegexProp()

havelaer committed:
improved tests touched files

Hypnosphi reviewed (approved):
LGTM. Let's merge it right after the upcoming `3.3.0` release

igor-dv commented on lib/ui/src/modules/ui/libs/hierarchy.js:
```diff
@@ -46,24 +46,56 @@ function fillHierarchy(namespaces, hierarchy, story) {
+  const rootNames = Object.keys(rootMap);
+  const hierarchies = rootNames.map(rootName => rootMap[rootName]);
+
+  // make sure the main root is the first
```
Please extract this block to function - `moveRootToFront` / `ensureRootIsFirst` or something

Havelaer replied:
Extracted to `ensureMainRootIsFirst` and used `Object.values` instead of `Object.keys(...).map(...)`

igor-dv reviewed (approved):
Very nice tests! + tested locally.

havelaer committed:
extracted main root is first of hierarchies

ndelangen commented:
Good to go if you ask me!

danielduan commented:
I think we're aiming to add this for `3.4`?

I added the `do not merge` label because of the upcoming changes to our workflow. Feel free to merge to master after we get `3.3` out

Hypnosphi commented:
We have moved some stories to `examples/storybook-official`. Would you mind resolving the conflicts?

havelaer committed:
post rebase origin/master fixes

havelaer committed:
update snapshots

havelaer committed:
updated cli snapshots

Hypnosphi commented:
@shilman after we merge this one, we won't be able to release `3.3` from `master` anymore

Havelaer commented:
I've rebased the branch on master, keeping the history nice & clean. Fixed tests, added separator config to examples/official-storybook and updated core and cli snapshots.

ndelangen commented:
@Havelaer That's fantastic work!

igor-dv commented:
@Havelaer maybe you can resolve the conflicts and merge?

ndelangen committed:
Merge branch 'master' into feature/multiple-hierarchies

# Conflicts:
#	examples/official-storybook/config.js
#	lib/cli/test/snapshots/angular-cli/package.json
#	lib/cli/test/snapshots/meteor/package.json
#	lib/cli/test/snapshots/react/package.json
#	lib/cli/test/snapshots/react_native/package.json
#	lib/cli/test/snapshots/react_native_scripts/package.json
#	lib/cli/test/snapshots/react_project/package.json
#	lib/cli/test/snapshots/react_scripts/package.json
#	lib/cli/test/snapshots/sfc_vue/package.json
#	lib/cli/test/snapshots/update_package_organisations/package.json
#	lib/cli/test/snapshots/vue/package.json
#	lib/cli/test/snapshots/webpack_react/package.json

ndelangen commented:
I resolved merge conflicts

igor-dv committed:
Merge branch 'master' into feature/multiple-hierarchies

igor-dv commented:
👏
